### ISSUE TITLE
Automatic renaming of duplicate class?

### ISSUE THREAD

**Issue by jsbain** [beeware/rubicon-objc#181]
in 

`_new_from_class_statement`

if a classname on the objc side already exists, the following is thrown:
    RuntimeError('An Objective-C class named {!r} already exists'.format(name))

While this behavior makes sense for code run from a command line, where the lifetime of new objc classes is short, in an environment like pythonista, where we only get one process, and so any created objccclasses can never be deleted, it would be nice to specify a flag to automatically rename the class on the objc side. 

Maybe some sort decorator on the class, which changes the class name (checks for existing name, and renames if needed ).  or an option that can be overridden in `api` after import.  obviously, one could never lookup the custom class by name with this scheme (alhough perhaps there could also be a dict of remapped classnames)

I am not familiar enough yet with how rubicon-objc is creating the class pair, and whether something like this is even possible.  Would it be possible to disassociate the objc class name from the python class name, or does rubicon rely on those always being the same?

**Comment by jsbain** [comment]
By auto renaming, I am specifically thinking of adding a numerical suffix to the base name (keep trying until a unique name is found) -- similar to the med used by pythonista when the debug flag is set in create_objc_class.

**Comment by dgelessus** [comment]
This would definitely be a useful feature for environments where restarting the Python interpreter is inconvenient (like Pythonista or a REPL).

As far as I can tell, it should be possible to implement this like how Pythonista does it in `objc_util`, i. e. automatically adjust the Objective-C name of the class until it doesn't conflict with an existing class anymore. rubicon.objc itself generally doesn't use the class name - it passes it to Objective-C when creating the class, but afterwards it only uses the `Class` pointer to work with the class (calling methods, etc.). So it *shouldn't* matter if the class has a different name than what the user requested.

Of course the same would also be required for the user code - that is, the user should only reference the newly created class via its Python name and cannot look it up by its Objective-C name via `ObjCClass`. If we make the behavior opt-in via a flag when the class is created, this shouldn't be an issue though - by enabling that flag, the user accepts that they are only allowed to use the class via its pointer/Python name and not via its Objective-C name. (Maybe we should even enforce that - if the user sets the "auto-rename" flag, *always* append an integer suffix, so that it's impossible to accidentally reference the class through a hardcoded Objective-C name.)

Regarding the syntax - Python 3 supports passing keyword arguments to `class` statements. We already use this when [defining protocols](https://rubicon-objc.readthedocs.io/en/latest/how-to/protocols.html#implementing-a-protocol) for example. I think this would work well here too, for example:

```python
class MySubclass(objc.NSObject, auto_rename=True):
    pass
```

(Let me know if you have any ideas for the name of the flag - `auto_rename` is just a suggestion/placeholder.)

**Comment by freakboy3742** [comment]
I agree this is a use case we should support.

I'm not sure a flag on the class definition is the right approach, though. Consider the following use case:
 * Start a REPL
 * Define a `TogaWindow` class
 * `import toga`

Importing Toga will cause a *new* `TogaWindow` class to be defined, which will raise an error. So - either we need to accept that user's can't define classes that won't be redefined by import modules, or we need to modify Toga to use auto-rename. The former doesn't sound like a great user experiences. The latter could inadvertently mask performance issues in production code.

An alternate approach is to make the auto-rename behavior a global runtime option of Rubicon; i.e., expose an `enable_auto_rename()` method that moves Rubicon from "strict" mode to auto-rename mode; if you're in a REPL context, you invoke that method at the start of your session, and then all class definitions are safe.

**Comment by dgelessus** [comment]
@freakboy3742 I understand what you mean - though the particular case you're describing is unlikely to happen in practice, because libraries generally use unique prefixes for their class names, specifically to avoid naming conflicts. A REPL user will probably use their own prefix or none at all, which makes conflicts with imported libraries unlikely.

The general problem you're describing can still happen though, for example when using `importlib.reload`, in which case the reloaded library will conflict with the old version of itself. For those cases a global flag would be useful. This should definitely be disabled by default though, to avoid masking actual name conflicts outside of interactive/debugging use.

I think the class-level flag also has its uses though. I remember that someone asked about defining an Objective-C classes inside a Python function, basically as a closure. At the moment this isn't possible (because calling the function more than once would lead to a class name conflict), but using a class-level `auto_rename` flag, this usage could be supported. You need to be careful with this obviously - you shouldn't use this to define thousands of Objective-C classes, there are better and more performant ways to implement this kind of closure behavior.

**Comment by freakboy3742** [comment]
The closure use case is definitely an interesting one. As you say, it's definitely a potential footgun if abused, but as long as we label it clearly, it seems like a use case we should allow for. Sounds like both options (global and per-class) may be necessary.

**Comment by qqfunc** [comment]
Although it looks like a 4 year old Issue, I wrote a program because I'm a Pythonista user and really want this feature.

(There might be a mistake because I don't understand the whole system of Rubicon-ObjC and am a beginner in projects like this.)

I have also considered about using [`objc_disposeClassPair`](https://developer.apple.com/documentation/objectivec/1418912-objc_disposeclasspair?language=objc), but that's an unsafe way, isn't it?

**Comment by freakboy3742** [comment]
I haven't actually tried your code; but from a quick inspection, it looks to be on the right track.

I can see how disposing of the ObjC Class could potentially be an option - the issue would be tracking when it would be safe to invoke the destroy. You'd need to be certain that the ObjC runtime wasn't retaining any instances or references to the class - information that won't necessarily be easily obtainable from the Python side.

**Comment by freakboy3742** [comment]
Resolved by #419.
